src/plugins/tune/tune_impl.cpp

```diff
@@ -0,0 +1,161 @@
+                break;
+            case Tune::SongElement::STYLE_NORMAL:
+                tune_str.append("MN");
+                break;
+            case Tune::SongElement::STYLE_STACCATO:
+                tune_str.append("MS");
+                break;
+            case Tune::SongElement::DURATION_1:
+                tune_str.append("L1");
+                last_duration = 1;
+                break;
+            case Tune::SongElement::DURATION_2:
+                tune_str.append("L2");
+                last_duration = 2;
+                break;
+            case Tune::SongElement::DURATION_4:
+                tune_str.append("L4");
+                last_duration = 4;
+                break;
+            case Tune::SongElement::DURATION_8:
+                tune_str.append("L8");
+                last_duration = 8;
+                break;
+            case Tune::SongElement::DURATION_16:
+                tune_str.append("L16");
+                last_duration = 16;
+                break;
+            case Tune::SongElement::DURATION_32:
+                tune_str.append("L32");
+                last_duration = 32;
+                break;
+            case Tune::SongElement::NOTE_A:
+                tune_str.append("A");
+                break;
+            case Tune::SongElement::NOTE_B:
+                tune_str.append("B");
+                break;
+            case Tune::SongElement::NOTE_C:
+                tune_str.append("C");
+                break;
+            case Tune::SongElement::NOTE_D:
+                tune_str.append("D");
+                break;
+            case Tune::SongElement::NOTE_E:
+                tune_str.append("E");
+                break;
+            case Tune::SongElement::NOTE_F:
+                tune_str.append("F");
+                break;
+            case Tune::SongElement::NOTE_G:
+                tune_str.append("G");
+                break;
+            case Tune::SongElement::NOTE_PAUSE:
+                tune_str.append("P" + std::to_string(last_duration));
+                break;
+            case Tune::SongElement::SHARP:
+                tune_str.append("+");
+                break;
+            case Tune::SongElement::FLAT:
+                tune_str.append("-");
+                break;
+            case Tune::SongElement::OCTAVE_UP:
+                tune_str.append(">");
+                break;
+            case Tune::SongElement::OCTAVE_DOWN:
+                tune_str.append("<");
+                break;
+            default:
+                break;
+        }
+    }
+
+    LogDebug() << "About to send tune: " << tune_str;
+
+    if (tune_str.size() > 230) {
+        report_tune_result(callback, Tune::Result::TUNE_TOO_LONG);
+        return;
+    }
+
+    std::string tune1_str;
+    std::string tune2_str;
+
+    if (tune_str.size() < 30) {
+        tune1_str = tune_str;
+        tune2_str = "";
+    } else {
+        tune1_str = tune_str.substr(0, 30);
+        tune2_str = tune_str.substr(30);
```

Comment:
I think so. I call `tune1_str.c_str()` when I build the message, and that should be null-terminated.

Reply:
Ok, unless it's too long. Or are you checking that?

Reply:
Yes, you are, but I think you need to check `> 229` otherwise there is no space for the 0 termination, right?

Reply:
I'm not sure. So I tried to do `substr(0, 29)` to leave space for the NUL in the first string, but then the second string is not played. As if "either you fill the space, or you NUL-terminate early". Could that be?

Reply:
The first one should not be 0 terminated, correct! If it is, it won't continue with the second one.

However, I would say it's safer to make sure the second one is properly 0 terminated.

Reply:
So we lose 1 char to make sure that the autopilot implementation does not crash? :thinking: :sweat_smile: 

Your call, I'll change it. But I believe it would be a vulnerability if the autopilot did not manage that correctly...

Reply:
That's true. Don't you worry:
https://github.com/PX4/Firmware/pull/13158/files#diff-d9c39530a9fff85e8b746162990ca263R1664